internal/pkg/cli/deploy/env.go

```diff
@@ -100,11 +136,123 @@ func NewEnvDeployer(in *NewEnvDeployerInput) (*envDeployer, error) {
 			Env:             in.Env,
 		},
 		newStackSerializer: func(in *deploy.CreateEnvironmentInput, lastForceUpdateID string, oldParams []*awscfn.Parameter) stackSerializer {
-			return stack.NewEnvConfigFromExistingStack(in, lastForceUpdateID, oldParams)
+			return cfnstack.NewEnvConfigFromExistingStack(in, lastForceUpdateID, oldParams)
+		},
+		envDescriber:      envDescriber,
+		envManagerSession: envManagerSession,
+		lbDescriber:       elbv2.New(envManagerSession),
+		newServiceStackDescriber: func(svc string) stackDescriber {
+			return stack.NewStackDescriber(cfnstack.NameForService(in.App.Name, in.Env.Name, svc), envManagerSession)
 		},
 	}, nil
 }
 
+// Validate returns an error if the environment manifest is incompatible with services and application configurations.
+func (d *envDeployer) Validate(mft *manifest.Environment, output io.Writer) error {
+	return d.validateCDN(mft, output)
+}
+
+func (d *envDeployer) validateCDN(mft *manifest.Environment, output io.Writer) error {
+	isManagedCDNEnabled := mft.CDNEnabled() && !mft.HasImportedPublicALBCerts() && d.app.Domain != ""
+	if isManagedCDNEnabled {
+		// With managed domain, if the customer isn't using `alias` the A-records are inserted in the service stack as each service domain is unique.
+		// However, when clients enable CloudFront, they would need to update all their existing records to now point to the distribution.
+		// Hence, we force users to use `alias` and let the records be written in the environment stack instead.
+		if err := d.envDescriber.ValidateCFServiceDomainAliases(); err != nil {
+			return err
+		}
+	}
+
+	if mft.CDNEnabled() && mft.CDNDoesTLSTermination() && (mft.HasImportedPublicALBCerts() || d.app.Domain != "") {
+		err := d.validateALBWorkloadsDontRedirect()
+		var redirErr *errEnvHasPublicServicesWithRedirect
+		switch {
+		case errors.As(err, &redirErr) && mft.ALBIngressRestrictedToCDN():
+			return err
+		case errors.As(err, &redirErr):
+			fmt.Fprintln(output, redirErr.Warning())
+		case err != nil:
+			return fmt.Errorf("can't enable TLS termination on CDN: %w", err)
```

Comment:
nit: can we drop the `can't`? how does the error read if we have just `enable TLS termination on CDN: %w` (this is from Uber's go guideline:
> When adding context to returned errors, keep the context succinct by avoiding phrases like "failed to", which state the obvious and pile up as the error percolates up through the stack:
> https://github.com/uber-go/guide/blob/master/style.md#errors

I'll leave this to your judgement!